# Wrangle & Analyze WeRateDogs Data

<hr>

Real-world data rarely comes clean. Using Python and its libraries, I will gather data from a variety of sources and in a variety of formats, assess its quality and tidiness, then clean it. This is called data wrangling. I will document my wrangling efforts in a Jupyter Notebook, plus showcase them through analyses and visualizations using Python.

The dataset that I will be wrangling (and analyzing and visualizing) is the tweet archive of Twitter user @dog_rates, also known as WeRateDogs. WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "they're good dogs Brent." WeRateDogs has over 4 million followers and has received international media coverage.

<p align="center">
  <img src="img/dog-rates-social.jpg" width="600">
</p>

The WeRateDogs Twitter archive contains basic tweet data for all 5000+ of their tweets, but not everything. One column the archive does contain though: each tweet's text, which I used to extract rating, dog name, and dog "stage" (i.e. doggo, floofer, pupper, and puppo) to make this Twitter archive "enhanced." Of the 5000+ tweets, I have filtered for tweets with ratings only (there are 2356).

<p align="center">
  <img src="img/data.png" width="1300">
</p>

Retweet count and favorite count are two of the notable column omissions. Fortunately, this additional data can be gathered from Twitter's API, which I will do.

In [3]:
# import libs
import numpy as np
import pandas as pd
import requests
import json
import matplotlib.pyplot as plt
from stop_words import get_stop_words

# pandas settings
pd.set_option('display.max_colwidth', -1)

## Gathering Data

Gather data from various sources and a variety of file formats.

<hr>

* [Enhanced Twitter Archive](#Enhanced-Twitter-Archive)  
* [Image Predictions File](#Image-Predictions-File)  
* [Twitter API File](#Twitter-API-File)

### Enhanced Twitter Archive

This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017.

In [4]:
# load twitter archive
twitter_arch = pd.read_csv("data/twitter-archive-enhanced.csv")
# use tweet id column as index
twitter_arch.set_index("tweet_id", inplace = True)
# display few lines
twitter_arch.head(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
892420643555336193 NaN                    NaN                    
892177421306343426 NaN                    NaN                    
891815181378084864 NaN                    NaN                    

                                    timestamp  \
tweet_id                                        
892420643555336193  2017-08-01 16:23:56 +0000   
892177421306343426  2017-08-01 00:17:27 +0000   
891815181378084864  2017-07-31 00:18:03 +0000   

                                                                                                source  \
tweet_id                                                                                                 
892420643555336193  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
892177421306343426  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
891815181378084864  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                                  text  \
tweet_id                                                                                                                                                                                 
892420643555336193  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
892177421306343426  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
891815181378084864  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
892420643555336193 NaN                  NaN                         
892177421306343426 NaN                  NaN                         
891815181378084864 NaN                  NaN                         

                   retweeted_status_timestamp  \
tweet_id                                        
892420643555336193  NaN                         
892177421306343426  NaN                         
891815181378084864  NaN                         

                                                                      expanded_urls  \
tweet_id                                                                              
892420643555336193  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
892177421306343426  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
891815181378084864  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

                    rating_numerator  rating_denominator     name doggo  \
tweet_id                                                                  
892420643555336193  13                10                  Phineas  None   
892177421306343426  13                10                  Tilly    None   
891815181378084864  12                10                  Archie   None   

                   floofer pupper puppo  
tweet_id                                 
892420643555336193  None    None   None  
892177421306343426  None    None   None  
891815181378084864  None    None   None

### Image Predictions File

This file contains top three predictions of dog breed for each dog image from the WeRateDogs archive. Table contains the top three predictions, tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images).

In [5]:
# get file with the image predictions
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
with open('data/image-predictions.tsv' , 'wb') as file:
    predictions = requests.get(url)
    file.write(predictions.content)

# load image predictions
image_pred = pd.read_csv('data/image-predictions.tsv', sep = '\t')
# use tweet id column as index
image_pred.set_index("tweet_id", inplace = True)
# display few lines
image_pred.head(3)

jpg_url  img_num  \
tweet_id                                                                       
666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg  1         
666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg  1         
666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg  1         

                                        p1   p1_conf  p1_dog  \
tweet_id                                                       
666020888022790149  Welsh_springer_spaniel  0.465074  True     
666029285002620928  redbone                 0.506826  True     
666033412701032449  German_shepherd         0.596461  True     

                                    p2   p2_conf  p2_dog                   p3  \
tweet_id                                                                        
666020888022790149  collie              0.156665  True    Shetland_sheepdog     
666029285002620928  miniature_pinscher  0.074192  True    Rhodesian_ridgeback   
666033412701032449  malinois            0.138584  True    bloodhound            

                     p3_conf  p3_dog  
tweet_id                              
666020888022790149  0.061428  True    
666029285002620928  0.072010  True    
666033412701032449  0.116197  True

### Twitter API File

Retweet count and favorite count are two of the notable column omissions of Twitter data archive. Fortunately, this additional data can be gathered from Twitter's API. Twitter API file contains tweet id, favorite count and retweet count. 

In [6]:
# load twitter API data
with open('data/tweet-json.txt') as f:
    twitter_api = pd.DataFrame((json.loads(line) for line in f), columns = ['id', 'favorite_count', 'retweet_count'])

# change column names
twitter_api.columns = ['tweet_id', 'favorites', 'retweets']
# use tweet id column as index
twitter_api.set_index('tweet_id', inplace = True)
# display few lines
twitter_api.head(3)

favorites  retweets
tweet_id                               
892420643555336193  39467      8853    
892177421306343426  33819      6514    
891815181378084864  25461      4328

## Assessing Data

Assess data visually and programmatically for quality and tidiness issues using pandas.

<hr>

* [Twitter Archive Data](#Twitter-Archive-Data)  
* [Image Predictions](#Image-Predictions)  
* [Twitter API Data](#Twitter-API-Data)

### Twitter Archive Data

In [7]:
# display sample of data
twitter_arch.sample(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
712438159032893441 NaN                    NaN                    
701889187134500865 NaN                    NaN                    
819952236453363712 NaN                    NaN                    

                                    timestamp  \
tweet_id                                        
712438159032893441  2016-03-23 00:37:48 +0000   
701889187134500865  2016-02-22 21:59:57 +0000   
819952236453363712  2017-01-13 17:00:21 +0000   

                                                                                                source  \
tweet_id                                                                                                 
712438159032893441  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
701889187134500865  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
819952236453363712  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                                      text  \
tweet_id                                                                                                                                                                                     
712438159032893441  This is Kane. He's a semi-submerged Haitian Huffleplop. Happy af. Sick waterfall. 11/10 would pat head approvingly https://t.co/7zjEC501Ul                               
701889187134500865  This is Bilbo. He's not emotionally prepared to enter the water. 11/10 don't struggle Bilbo https://t.co/rH9SQgZUnQ                                                      
819952236453363712  This is Oliver. He has dreams of being a service puppo so he can help his owner. 13/10 selfless af\n\nmake it happen:\nhttps://t.co/f5WMsx0a9K https://t.co/6lJz0DKZIb   

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
712438159032893441 NaN                  NaN                         
701889187134500865 NaN                  NaN                         
819952236453363712 NaN                  NaN                         

                   retweeted_status_timestamp  \
tweet_id                                        
712438159032893441  NaN                         
701889187134500865  NaN                         
819952236453363712  NaN                         

                                                                                                                expanded_urls  \
tweet_id                                                                                                                        
712438159032893441  https://twitter.com/dog_rates/status/712438159032893441/photo/1                                             
701889187134500865  https://twitter.com/dog_rates/status/701889187134500865/photo/1                                             
819952236453363712  https://www.gofundme.com/servicedogoliver,https://twitter.com/dog_rates/status/819952236453363712/photo/1   

                    rating_numerator  rating_denominator    name doggo  \
tweet_id                                                                 
712438159032893441  11                10                  Kane    None   
701889187134500865  11                10                  Bilbo   None   
819952236453363712  13                10                  Oliver  None   

                   floofer pupper  puppo  
tweet_id                                  
712438159032893441  None    None   None   
701889187134500865  None    None   None   
819952236453363712  None    None   puppo

In [8]:
# pint a summary of a DataFrame
twitter_arch.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 892420643555336193 to 666020888022790149
Data columns (total 16 columns):
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), object(10)
memory usa

In [9]:
# check if ids are unique
twitter_arch.index.is_unique

True

In [10]:
# check number of replies
np.isfinite(twitter_arch.in_reply_to_status_id).sum()

78

In [11]:
# check values in sources
twitter_arch.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                        91  
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                     33  
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>    11  
Name: source, dtype: int64

In [12]:
# check quality of text
twitter_arch.text.sample(3)

tweet_id
733822306246479872    This is Moose. He's a Polynesian Floofer. Dapper af. 10/10 would pet diligently https://t.co/mVfqRdppTL                         
669993076832759809    This is Zoey. Her dreams of becoming a hippo ballerina don't look promising. 9/10 it'll be ok puppers https://t.co/kR1fqy4NKK   
666102155909144576    Oh my. Here you are seeing an Adobe Setter giving birth to twins!!! The world is an amazing place. 11/10 https://t.co/11LvqN4WLq
Name: text, dtype: object

In [13]:
# check number of retweets
np.isfinite(twitter_arch.retweeted_status_id).sum()

181

In [14]:
# check expanded urls
twitter_arch[~twitter_arch.expanded_urls.str.startswith(('https://twitter.com','http://twitter.com', 'https://vine.co'), na=False)].sample(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
856288084350160898  8.562860e+17           2.792810e+08          
855860136149123072  8.558585e+17           1.361572e+07          
848213670039564288  8.482121e+17           4.196984e+09          

                                    timestamp  \
tweet_id                                        
856288084350160898  2017-04-23 23:26:03 +0000   
855860136149123072  2017-04-22 19:05:32 +0000   
848213670039564288  2017-04-01 16:41:12 +0000   

                                                                                                source  \
tweet_id                                                                                                 
856288084350160898  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
855860136149123072  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
848213670039564288  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                                     text  \
tweet_id                                                                                                                                                                                    
856288084350160898  @xianmcguire @Jenna_Marbles Kardashians wouldn't be famous if as a society we didn't place enormous value on what they do. The dogs are very deserving of their 14/10   
855860136149123072  @s8n You tried very hard to portray this good boy as not so good, but you have ultimately failed. His goodness shines through. 666/10                                   
848213670039564288  Jerry just apuppologized to me. He said there was no ill-intent to the slippage. I overreacted I admit. Pupgraded to an 11/10 would pet                                 

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
856288084350160898 NaN                  NaN                         
855860136149123072 NaN                  NaN                         
848213670039564288 NaN                  NaN                         

                   retweeted_status_timestamp expanded_urls  rating_numerator  \
tweet_id                                                                        
856288084350160898  NaN                        NaN           14                 
855860136149123072  NaN                        NaN           666                
848213670039564288  NaN                        NaN           11                 

                    rating_denominator  name doggo floofer pupper puppo  
tweet_id                                                                 
856288084350160898  10                  None  None  None    None   None  
855860136149123072  10                  None  None  None    None   None  
848213670039564288  10                  None  None  None    None   None

In [15]:
# check for two or more urls in the expanded urls
twitter_arch[twitter_arch.expanded_urls.str.contains(',', na=False)].expanded_urls.count()

639

In [16]:
# check rating denominator
twitter_arch.rating_denominator.value_counts()

10     2333
11     3   
50     3   
80     2   
20     2   
2      1   
16     1   
40     1   
70     1   
15     1   
90     1   
110    1   
120    1   
130    1   
150    1   
170    1   
7      1   
0      1   
Name: rating_denominator, dtype: int64

In [17]:
# check rating numerator
twitter_arch.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
420     2  
0       2  
15      2  
75      2  
80      1  
20      1  
24      1  
26      1  
44      1  
50      1  
60      1  
165     1  
84      1  
88      1  
144     1  
182     1  
143     1  
666     1  
960     1  
1776    1  
17      1  
27      1  
45      1  
99      1  
121     1  
204     1  
Name: rating_numerator, dtype: int64

In [18]:
# check for any float ratings in the text column
twitter_arch[twitter_arch.text.str.contains(r'\d+\.\d+\/\d+')][['text','rating_denominator', 'rating_numerator']]

text  \
tweet_id                                                                                                                                                                                
883482846933004288  This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948                            
832215909146226688  RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…                        
786709082849828864  This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                    
778027034220126208  This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
681340665377193984  I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace                                                          
680494726643068929  Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD                                                          

                    rating_denominator  rating_numerator  
tweet_id                                                  
883482846933004288  10                  5                 
832215909146226688  10                  75                
786709082849828864  10                  75                
778027034220126208  10                  27                
681340665377193984  10                  5                 
680494726643068929  10                  26

In [19]:
# check name of dog
twitter_arch.name.value_counts()

None        745
a           55 
Charlie     12 
Cooper      11 
Lucy        11 
Oliver      11 
Lola        10 
Penny       10 
Tucker      10 
Winston     9  
Bo          9  
Sadie       8  
the         8  
Buddy       7  
Daisy       7  
Bailey      7  
Toby        7  
an          7  
Leo         6  
Oscar       6  
Milo        6  
Rusty       6  
Bella       6  
Stanley     6  
Jack        6  
Koda        6  
Jax         6  
Scout       6  
Dave        6  
Bentley     5  
           ..  
Evy         1  
Rumpole     1  
Sky         1  
Tobi        1  
Newt        1  
Carbon      1  
Harlso      1  
Carll       1  
Laika       1  
Opie        1  
Blue        1  
life        1  
Lucia       1  
Jennifur    1  
Lipton      1  
Ruffles     1  
Brady       1  
Andy        1  
Emma        1  
Puff        1  
Steve       1  
Kramer      1  
Apollo      1  
Arlen       1  
Zeek        1  
Keet        1  
Champ       1  
Billy       1  
Benny       1  
Divine      1  
Name: name, Length: 957,

In [20]:
# check for stop words in dog name
# https://stackoverflow.com/a/5486535/7382214

stop_words = set(get_stop_words('en'))

count=0
for word in twitter_arch.name:
    if word.lower() in stop_words:
        count += 1
print('Rows with stop words:', count)

Rows with stop words: 83


In [21]:
# check if gods have more than one category assigned
categories = ['doggo', 'floofer', 'pupper', 'puppo']

for category in categories:
    twitter_arch[category] = twitter_arch[category].apply(lambda x: 0 if x=='None' else 1)

twitter_arch['number_categories'] = twitter_arch.iloc[:,[12,13,14,15]].sum(axis=1)

In [22]:
# dogs categories
twitter_arch['number_categories'].value_counts()

0    1976
1    366 
2    14  
Name: number_categories, dtype: int64

#### Quality & Tidiness Issues

- in_reply_to_status_id, in_reply_to_user_id have an incorrect datatype - is a float, should be int;
- some of the gathered tweets are replies;
- the timestamp has an incorrect datatype - is an object, should be DateTime;
- source is an HTML element - it's text should be extracted;
- some rows in the text column begin from 'RT @dog_rates:';
- some rows in the text column have leading and/or trailing whitespace;
- some of the gathered tweets are retweets;
- retweeted_status_id, retweeted_status_user_id have an incorrect datatype - is a float, should be int;
- the retweeted timestamp has an incorrect datatype - is an object, should be DateTime;
- we have 59 missing expanded urls;
- we have 639 expanded urls which contain more than one url address;
- denominator of some ratings is not 10;
- numerator of some ratings is almost always greater than 10;
- float ratings have been incorrectly read from the text of tweet;
- 'None' in the name should be convert to NaN;
- we have stop words in the name column;
- dog 'stage' classification (doggo, floofer, pupper or puppo) should be one column;
- some dogs have more than one category assigned;

### Image Predictions

In [23]:
# display sample of data
image_pred.sample(3)

jpg_url  img_num  \
tweet_id                                                                       
724004602748780546  https://pbs.twimg.com/media/CgwuWCeW4AAsgbD.jpg  3         
670069087419133954  https://pbs.twimg.com/media/CUyQRzHWoAAhF1D.jpg  1         
858471635011153920  https://pbs.twimg.com/media/C-nnZBdXkAAB-wg.jpg  1         

                             p1   p1_conf  p1_dog         p2   p2_conf  \
tweet_id                                                                 
724004602748780546  Siamese_cat  0.950526  False   pug        0.018877   
670069087419133954  boathouse    0.313829  False   birdhouse  0.138331   
858471635011153920  Pembroke     0.987407  True    Cardigan   0.008723   

                    p2_dog       p3   p3_conf  p3_dog  
tweet_id                                               
724004602748780546  True    quilt    0.007628  False   
670069087419133954  False   ashcan   0.045673  False   
858471635011153920  True    basenji  0.003424  True

In [24]:
# pint a summary of a DataFrame
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 666020888022790149 to 892420643555336193
Data columns (total 11 columns):
jpg_url    2075 non-null object
img_num    2075 non-null int64
p1         2075 non-null object
p1_conf    2075 non-null float64
p1_dog     2075 non-null bool
p2         2075 non-null object
p2_conf    2075 non-null float64
p2_dog     2075 non-null bool
p3         2075 non-null object
p3_conf    2075 non-null float64
p3_dog     2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(4)
memory usage: 152.0+ KB


In [25]:
# check if ids are unique
image_pred.index.is_unique

True

In [26]:
# check jpg_url
image_pred[~image_pred.jpg_url.str.endswith(('.jpg', '.png'), na=False)].jpg_url.count()

0

In [27]:
# check image number
image_pred.img_num.value_counts()

1    1780
2    198 
3    66  
4    31  
Name: img_num, dtype: int64

In [28]:
# check 1st prediction
image_pred.p1.sample(3)

tweet_id
670361874861563904    platypus
669753178989142016    Pembroke
666996132027977728    hay     
Name: p1, dtype: object

In [29]:
# check dog predictions
image_pred.p1_dog.count()

2075

#### Quality & Tidiness Issues

- the dataset has 2075 entries, while twitter archive dataset has 2356 entries;
- column names are confusing and do not give much information about the content;
- dog breeds contain underscores, and have different case formatting;
- only 2075 images have been classified as dog images for top prediction;
- dataset should be merged with the twitter archive dataset;

### Twitter API Data

In [30]:
# display sample of data
twitter_api.sample(3)

favorites  retweets
tweet_id                               
703774238772166656  2020       526     
673662677122719744  1339       397     
667455448082227200  203        66

In [31]:
# pint a summary of a DataFrame
twitter_api.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2354 entries, 892420643555336193 to 666020888022790149
Data columns (total 2 columns):
favorites    2354 non-null int64
retweets     2354 non-null int64
dtypes: int64(2)
memory usage: 55.2 KB


In [32]:
# check if ids are unique
twitter_arch.index.is_unique

True

#### Quality & Tidiness Issues

- twitter archive dataset has 2356 entries, while twitter API data has 2354;
- dataset should be merged with the twitter archive dataset;

## Cleaning Data

Using pandas, clean the quality and tidiness issues identified in the [Assessing Data](#Assessing-Data) section.

<hr>

* [Twitter Archive Data](#Twitter-Archive-Data)  
* [Image Predictions](#Image-Predictions)  
* [Twitter API Data](#Twitter-API-Data)

## Analyze Data

Analyze and visualize data using matplotlib.

<hr>

* [Twitter Archive Data](#Twitter-Archive-Data)  
* [Image Predictions](#Image-Predictions)  
* [Twitter API Data](#Twitter-API-Data)